Unit test failures on 2.13.12

Unit tests fail on the Scala 2 community build.

## Configuration (required) ##

HEAD

## Problem

```scala
[scalafmt] [error] Failed tests:
[scalafmt] [error] 	org.scalafmt.FidelityTest
[scalafmt] [error] 	org.scalafmt.FormatTests
```

## Notes

It looks like `Tokens#last` doesn't throw on empty `Tokens`. Scalafmt should handle empty tokens:
```
defnBeforeTemplate(leftOwner).filter(_.tokens.nonEmpty).map { x =>
  val policyEnd = Policy.End.On(x.tokens.last)
  delayedBreakPolicy(policyEnd)(forceNewlineBeforeExtends)
}
```
(I didn't pursue a fix.)

Noticed at https://github.com/scala/community-build/issues/1680#issuecomment-1625131437

```
==> X org.scalafmt.FormatTests.unit/Annotations.stat:59: #1640                                       |  0.001s org.scalafmt.util.FormatException: java.util.NoSuchElementException: last of empty IndexedSeq
    at org.scalafmt.FormatTests.run(FormatTests.scala:56)
    at org.scalafmt.FormatTests.$anonfun$new$3(FormatTests.scala:33)
    at org.scalafmt.FormatTests.$anonfun$new$3$adapted(FormatTests.scala:33)
    at org.scalafmt.util.CanRunTests.$anonfun$runTest$2(CanRunTests.scala:19)
Caused by: java.util.NoSuchElementException: last of empty IndexedSeq
    at scala.collection.IndexedSeqOps.last(IndexedSeq.scala:110)
    at scala.collection.IndexedSeqOps.last$(IndexedSeq.scala:105)
    at scala.meta.tokens.Tokens.last(Tokens.scala:29)
    at org.scalafmt.internal.Router.$anonfun$getSplitsImpl$118(Router.scala:838)
    at scala.Option.map(Option.scala:242)
    at org.scalafmt.internal.Router.getSplitsImpl(Router.scala:837)
```

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
@@ -374,6 +374,13 @@ class FormatOps(
     new Policy.Delay(policy, Policy.End.Before(from))
   }
 
+  val WithTemplateOnLeft = new ExtractFromMeta(ft =>
+    ft.meta.leftOwner match {
+      case lo: Stat.WithTemplate => Some(lo.templ)
+      case _ => None
+    }
+  )
+
   def templateCurlyFt(template: Template): Option[FormatToken] =
     getStartOfTemplateBody(template).map(tokenBefore)
 

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -817,14 +817,10 @@ class Router(formatOps: FormatOps) {
       case FormatToken(
             _: T.KwObject | _: T.KwClass | _: T.KwTrait | _: T.KwEnum,
             _,
-            _
+            WithTemplateOnLeft(template)
           ) =>
-        def expire = defnTemplate(leftOwner)
-          .flatMap {
-            getTemplateGroups(_).flatMap(
-              _.lastOption.flatMap(_.headOption.flatMap(_.tokens.headOption))
-            )
-          }
+        def expire = getTemplateGroups(template)
+          .flatMap(_.last.headOption.flatMap(_.tokens.headOption))
           .getOrElse(getLastToken(leftOwner))
         def forceNewlineBeforeExtends = Policy.before(expire) {
           case Decision(FormatToken(_, soft.ExtendsOrDerives(), m), s)
@@ -834,8 +830,8 @@ class Router(formatOps: FormatOps) {
         val policy =
           if (style.binPack.keepParentConstructors) None
           else
-            defnBeforeTemplate(leftOwner).map { x =>
-              val policyEnd = Policy.End.On(x.tokens.last)
+            template.tokens.headOption.map { head =>
+              val policyEnd = Policy.End.Before(head)
               delayedBreakPolicy(policyEnd)(forceNewlineBeforeExtends)
             }
         Seq(Split(Space, 0).withPolicyOpt(policy))

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenOps.scala
@@ -123,16 +123,6 @@ object TokenOps {
       case _ => None
     }
 
-  def defnBeforeTemplate(tree: Tree): Option[Tree] =
-    tree match {
-      case t: Defn.Object => Some(t.name)
-      case t: Defn.Class => Some(t.ctor)
-      case t: Defn.Trait => Some(t.ctor)
-      case t: Defn.Enum => Some(t.ctor)
-      case t: Pkg.Object => Some(t.name)
-      case _ => None
-    }
-
   val formatOnCode = Set(
     "@formatter:on", // IntelliJ
     "format: on" // scalariform

